crates/engine/src/test_api.rs

```diff
@@ -0,0 +1,293 @@
+// Copyright 2018-2021 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::{
+    ext::Engine,
+    types::{
+        AccountId,
+        Balance,
+    },
+    AccountError,
+    Error,
+};
+use std::collections::HashMap;
+
+/// Record for an emitted event.
+#[derive(Clone)]
+pub struct EmittedEvent {
+    /// Recorded topics of the emitted event.
+    pub topics: Vec<Vec<u8>>,
+    /// Recorded encoding of the emitted event.
+    pub data: Vec<u8>,
+}
+
+#[derive(Clone)]
+pub struct RecordedPrintlns {
+    printlns: Vec<String>,
+}
+
+impl RecordedPrintlns {
+    // Creates a new `Engine instance.
+    pub fn new() -> Self {
+        Self {
+            printlns: Vec::new(),
+        }
+    }
+
+    // Records a new println.
+    pub fn record(&mut self, println: String) {
+        self.printlns.push(println);
+    }
+
+    // Clears all recorded printlns.
+    pub fn clear(&mut self) {
+        self.printlns.clear();
+    }
+}
+
+impl Default for RecordedPrintlns {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl IntoIterator for RecordedPrintlns {
+    type Item = String;
+    type IntoIter = std::vec::IntoIter<Self::Item>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.printlns.into_iter()
+    }
+}
+
+/// Recorder for relevant interactions with this crate.
+pub struct DebugInfo {
+    /// Emitted events recorder.
+    emitted_events: Vec<EmittedEvent>,
+    /// Emitted print messages recorder.
+    emitted_printlns: RecordedPrintlns,
+    /// The total number of reads to the storage.
+    count_reads: HashMap<AccountId, usize>,
+    /// The total number of writes to the storage.
+    count_writes: HashMap<AccountId, usize>,
+    /// The number of storage cells used by each account id.
+    cells_per_account: HashMap<AccountId, HashMap<Vec<u8>, bool>>,
```

Comment:
I think it would be nice to have additional `count_reads` and `count_writes` over `HashMap<Vec<u8>, usize>` that mirrors the number of read and writes to every storage database cell agnostic to an account ID. Multiple contracts may share some storage space through runtime chain extensions etc. or balance changes etc.

Reply:
Created a number of follow-ups now.

Reply:
Done. Closed.